### DISCUSSION TITLE
How to turn off SSR when using SPA mode

### DISCUSSION THREAD

**Discussion by garth** [remix-run/remix#8934] [Q&A]
For some reason remix dev is still doing SSR when running in SPA mode. Meaning that you get errors when using client only features.

Is there a way to turn off SSR when writing SPAs?

**Comment by sergiodxa** [comment]
Since your code needs to run to generate the initial `index.html`, you code should still not depend on client-only features on parts of the app that will run at build-time.

**Reply by garth** [reply]
Interesting.  So I guess remix SPA is not really suitable for SPAs where you would normally expect logic to only be executed in a browser context.

Is this someting that could get fixed in future releases or a limitation of being based on remix?

**Reply by ryan-zayne** [reply]
I've been pondering on this for a while and it really makes me wonder why it's called SPA mode when it's in fact not exactly?

**Reply by sergiodxa** [reply]
Your app runs client side, it just needs to run outside a browser one at build time, and only to generate the index.html, after that is an SPA.

**Reply by brophdawg11** [reply]
Can someone provide a minimal reproduction of an issue?  That will make it easier to understand the specific problem.  

The only thing that renders on the server is the root route's `HydrateFallback` - so as long as it's not using browser-only features you should be ok.  Other times if it's deps with browser-only side effects, the answer is to load them in `entry.client.tsx` so they don't even show up in the server build.

> So I guess remix SPA is not really suitable for SPAs where you would normally expect logic to only be executed in a browser context.

It should be - all of your loaders only ever execute on the client, and all components below the root only ever execute on the client

> why it's called SPA mode when it's in fact not exactly

Your app is absolutely in "SPA Mode" and runs as a SPA deployed to a static CDN via an `index.html` file.  I don't think the fact that there is a build process running on a node server disqualifies that description :)

**Reply by yesmeck** [reply]
Currently, we are having trouble making `@mui/x-charts` work with the SPA mode. Here is the minimal reproduction: https://github.com/yesmeck/remix-spa-issue

**Reply by garth** [reply]
React Router + Vite works fine for creating SPA apps. RemixSPA does not.  Really hoping that the merger announcement doesn't break things.

**Reply by brophdawg11** [reply]
@yesmeck I think this is a problem with MUI not being properly configured for ESM/CJS specifically, not Remix SPA mode (I also get similar errors when I turn off `ssr:false`).  MUI has historically been a source of pain when it comes to ESM/CJS (https://github.com/remix-run/remix/issues/8828).

I didn't dig too deep but using the config from that issue, I can build and run your app in SPA mode - but I still run into ESM/CJS issues on with Vite in dev mode:

```ts
export default defineConfig({
  plugins: [
    remix({ ssr: false }),
    tsconfigPaths(),
    ...(process.env.NODE_ENVIRONMENT === "development"
      ? [
          cjsInterop({
            dependencies: ["@mui/*"],
          }),
        ]
      : []),
  ],
  ssr: {
    noExternal: ["@mui/*"],
  },
});
```

**Reply by brophdawg11** [reply]
@garth or @Ryan-Zayne would either of you be able to provide a small reproduction of the issue you're having with browser APIs causing errors during the Remix SPA mode build?

**Reply by garth** [reply]
The problem is not browser api's it's that you might be creating apps for other environments, such as socketsupply, where you need to import something like this:

```ts
import { network } from 'socket:network'
```

With ReactRouter + Vite you can configure it to ignore specific imports, but last time I tried, this was not possible with Remix + Vite.

**Reply by yesmeck** [reply]
@brophdawg11 Yeah, it's an issue with MUI indeed. The thing is, as @garth pointed out, it works with React Router + Vite without extra configurations. What I'm expecting is it can work with Remix SPA without extra configurations as well. But now, just because of the SSR of HydrateFallback, I need to deal with these SSR and ESM/CJS issues.

**Comment by yesmeck** [comment]
This is the biggest blocker to migrate our large react-router app to Remix SPA mode. Our dependencies are mixed with CJS and ESM. Making them work together with SSR is quite difficult.

**Reply by yesmeck** [reply]
I patched the remix vite plugin to let the server always render an empty component. It seems to works well for me.

```diff
diff --git a/node_modules/@remix-run/dev/dist/vite/plugin.js b/node_modules/@remix-run/dev/dist/vite/plugin.js
index 5fee221..a97b830 100644
--- a/node_modules/@remix-run/dev/dist/vite/plugin.js
+++ b/node_modules/@remix-run/dev/dist/vite/plugin.js
@@ -438,6 +438,15 @@ const remixVitePlugin = (remixUserConfig = {}) => {
     ctx.serverBundleBuildConfig.routes :
     // Otherwise, all routes are imported as usual
     ctx.remixConfig.routes;
+
+    routes = Object.keys(routes).reduce((acc, id) => {
+      acc[id] = {
+        ...routes[id],
+        file: 'root.server.tsx'
+      };
+      return acc;
+    }, {});
+
     return `
     import * as entryServer from ${JSON.stringify(resolveFileUrl.resolveFileUrl(ctx, ctx.entryServerFilePath))};
     ${Object.keys(routes).map((key, index) => {
```

```ts
// routes/root.server.tsx
import { Links, Meta, Scripts, ScrollRestoration } from '@remix-run/react';

export function Layout({ children }: { children: React.ReactNode }) {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}

export default function App() {
  return null;
}

export function HydrateFallback() {
  return null;
}
```
